### ISSUE TITLE
Define task skipping behavior

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#9125]
Gradle has different strategies to completely "skip" the execution of tasks, and there are some slight discrepancies in how we handle them wrt:

* we keep empty `@OutputDirectory`s from the previous (non-skipped) execution for some ("no action", `onlyIf()`), but remove them for others (`@SkipWhenEmpty`)
* whether we clean up stale output files from the previous (non-skipped) execution or not; we do this for `@SkipWhenEmpty`, but don't for the others
* how we store the execution state; we used to not store it for any of them, but since 5.1 we explicitly remove it for `@SkipWhenEmpty`, and leave for the rest
* what does it mean to "skip" the task from the user's perspective? Skipping for "no actions" or with `@SkipWhenEmpty` is just a performance optimization, resulting in (roughly) equal output as if the task was truly executed, but `onlyIf()` is weird, because now we leave the results from the previous execution in place

**Comment by wolfs** [comment]
Let's keep this open.

**Comment by walkerburgin** [comment]
> we keep empty @OutputDirectorys from the previous (non-skipped) execution for some ("no action", onlyIf()), but remove them for others (@SkipWhenEmpty)

Is this still the intended behavior?

We've run into trouble recently converting tasks to use `@SkipWhenEmpty`, and looking at this issue I'm not sure whether it's a bug, we're using it wrong, or the behavior is just not what I expected. I set up a small repro here: https://github.com/walkerburgin/gradle-skip-when-empty. 

We have a task whose primary input is annotated with `@SkipWhenEmpty`. I can run the task and generate output files into the `build` directory: 

```
> tree build
build
└── typescript
    └── index.js
```

If I delete the source files and re-run the task with `--info` I get a log entry like this: 

```
> Task :compileTypeScript NO-SOURCE
Skipping task ':compileTypeScript' as it has no source files and no previous output files.
:compileTypeScript (Thread[Execution wo
```

The task was skipped like I wanted, but the files still exist in the `build` directory. It also says "no previous output files", which doesn't seem right.

```
> tree build
build
└── typescript
    └── index.js
```

In our real build, this causes cache misses for expensive downstream tasks (and it's subtle). I can go back to having the `@TaskAction` short circuit manually after deleting the output directory if there are no primary input files, but the clarity in build scans of `NO-SOURCE` is really nice.

**Comment by wolfs** [comment]
> > we keep empty @OutputDirectorys from the previous (non-skipped) execution for some ("no action", onlyIf()), but remove them for others (@SkipWhenEmpty)
> 
> Is this still the intended behavior?
> 

Yes, this is still intended behavior. There is a caveat though: Gradle only deletes the output if it is safe to delete. When the `base` plugin is applied, then everything in the `build` directory is deemed safe to delete. In your reproducer, if you apply the `base` plugin, then you can see the output files being deleted.

**Comment by walkerburgin** [comment]
Oh interesting! You're right that if I apply the base plugin the repro behaves the way that's described. That would explain it -- for historical reasons we don't apply the `base` plugin (yet) 😬. Is there any other way to signal that a task's outputs are safe to delete? If not I guess it's another good reason to prioritize applying the `base` plugin.

**Comment by wolfs** [comment]
> Is there any other way to signal that a task's outputs are safe to delete?

Currently, there isn't a public API for this. So you'd need to apply the `base` plugin. What is the reason why you don't apply it?

**Comment by walkerburgin** [comment]
Sounds good, thank you!

We're using Gradle to build several large (thousands of projects each) frontend monorepos. When we migrated to Gradle a couple years ago from raw NPM scripts we prioritized matching the old semantics of lifecycle tasks for developers so that `npm run build` translated directly to `./gradlew build` (etc). Very few of our frontend devs had used Gradle in other contexts. If I were going to do it over again I'd bite the bullet and adopt the more idiomatic gradle base plugin & lifecycle from day one.
